02-KernelModules_CharacterDevice/chrdev_driver.c

```diff
@@ -0,0 +1,188 @@
+#include <linux/slab.h>			/* kfree(), kmalloc() */
+#include <linux/fs.h>           /* file_operations */
+#include <linux/types.h>        /* dev_t */
+#include <linux/cdev.h>
+#include <asm/uaccess.h>        /* copy_*_user */
+
+#include "chrdev_driver.h"
+
+#ifndef MODULE_NAME
+#define MODULE_NAME "chrdev_driver"
+#endif
+
+#define MODULE_MIN_BUFF_SIZE 1024
+
+int chrdev_major = 0;		/* MAJOR */
+int chrdev_minor = 0;		/* MINOR */
+int chrdev_nr_devs = 1;		/* Number device for register */
+
+unsigned int chrdev_buffer_size = 0;	
+module_param(chrdev_buffer_size, uint, S_IRUGO);
+
+struct cdev *cdev;
+
+static char *chrdev_buffer;
+
+/*
+ * Open device
+ */
+int chrdev_open(struct inode *inode, struct file *filp)
+{
+	printk(KERN_INFO " chrdev: success open \n");
+	return 0;
+}
+
+/*
+ * Free device
+ */
+int chrdev_release(struct inode *inode, struct file *filp)
+{
+	printk(KERN_INFO " chrdev: success close \n");
+	return 0;
+}
+
+/*
+ * Read device
+ */
+ssize_t chrdev_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
+{
+	// return cmd
+	ssize_t retval = 0;
+	
+	// number of bytes left to read
+	int remain = chrdev_buffer_size - (int) (*f_pos); 
+	
+	if (remain == 0) 
+		return 0;	
+	if (count > remain)
+		count = remain;
+
+	if (raw_copy_from_user(buf, chrdev_buffer + *f_pos, count)) {
+		retval = -EFAULT;
+		goto out;
+	}
+	*f_pos += count;
+	retval = count;
+
+  out:
+return retval;
+}
+
+/*
+ * Write device
+ */
+ssize_t chrdev_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos)
+{
+	// return cmd
+	ssize_t retval = 0;
+	
+	// number of bytes left to write
+	int remain = chrdev_buffer_size - (int) (*f_pos); 
+
+	if (count > remain) {
+		return -EIO;
+	}
+
+	if (raw_copy_from_user(chrdev_buffer + *f_pos, buf, count)) {
+		retval = -EFAULT;
+		goto out;
+	}
+	
+	*f_pos += count;
+	retval = count;
+
+  out:
+return retval;
+}
+
+/*
+ * Driver for work with device
+ */
+struct file_operations chrdev_fops = {
+	.owner =    THIS_MODULE,
+	.read =     chrdev_read,
+	.write =    chrdev_write,
+	.open =     chrdev_open,
+	.release =  chrdev_release,
+};
+
+/*
+ * Free device and module
+ */
+void __exit chrdev_exit(void)
+{
+	dev_t devno = MKDEV(chrdev_major, chrdev_minor);
+	if (cdev) {
+		cdev_del(cdev);
+	}
+
+	if (chrdev_buffer) 
+		kfree(chrdev_buffer);
+ 	
+	unregister_chrdev_region(devno, chrdev_nr_devs);
+
+	printk(KERN_INFO "chrdev: exit \n");
+}
+
+/*
+ * Init module and driver
+ */
+int __init chrdev_init(void)
+{
+	int result = 0;
+	dev_t dev = 0;
+
+	result = alloc_chrdev_region(&dev, chrdev_minor, chrdev_nr_devs, MODULE_NAME);
+	chrdev_major = MAJOR(dev);
+
+	if (result) {
+		printk(KERN_WARNING "chrdev: can't get major %d\n", chrdev_major);
+		return result;
+	}
+
+	cdev = cdev_alloc();
+ 	
+	 if (!cdev) {
+		result = -ENOMEM;
+		goto fail; 
+	}
+	cdev_init(cdev, &chrdev_fops);
+	cdev->owner   = THIS_MODULE;
+	if (cdev_add(cdev, dev, chrdev_nr_devs)) { 
```

Comment:
```suggestion
	if (cdev_add(cdev, dev, chrdev_nr_devs)) { 
			printk(KERN_WARNING "chrdev: cdev_add error \n");
		}
		printk(KERN_INFO "chrdev: %d:%d \n", chrdev_major, chrdev_minor);
	
```

printk(KERN_WARNING  "chrdev: cdev_add error \n");

Please do it error handling  for this.

Think of fatal errors and their handling.